### PR TITLE
[CRASH] Scan segfault with inconsistent inner graphs

### PR DESCRIPTION
**Pull request by carriepl**
This PR adds some validation and a test case for the recent segfault issue when the inner graph has inconsistencies. The validation ensures that, for recurrent states, the inner inputs and inner outputs have the same type and that there are no GPU variables in the inputs and/or outputs is the flags in scan_op.info are not set correctly.

This also includes some refactoring of some helper functions in Scan_op to travel between inner and outer variables as well as between inputs and outputs.

### CODE REVIEW

File: theano/scan_module/tests/test_scan.py [hunk 1]
```diff
@@ -1334,6 +1334,20 @@ def test_simple_shared_random(self):
...
+
+        # The pickled scan op used in this test requires the use of a gpu
+        from theano.sandbox import cuda
+        if not cuda.cuda_available:
+            raise SkipTest('Optional package cuda disabled')
+
+        # When unpickled, the scan op should perform validation on its inner
+        # graph, detect the inconsistencies and raise a TypeError
+        assert_raises(TypeError, cPickle.load,
+                      open("./inconsistent_scan.pkl", "r"))
```
Comment (nouiz) [inline theano/scan_module/tests/test_scan.py]: The file path isn't good. The current dir can be any dir. It isn't the directory where the tests is.

**Comment by nouiz** [comment]
I think the test will fail with python 3 due to the pickle. Can you test it with python 3 on a computer with a GPU? I'm fine to skip the test if there is other unpickle problem in the graph.

**Comment by carriepl** [comment]
I tested it with numpy 1.9.2 on python 3.4 and it works but I did not try with previous versions. I don't know if a different numpy/python version would break it though.

**Comment by nouiz** [comment]
Can you try with numpy 1.8? It is possible that older version do not work
correctly. Just skip the test if that happen with that version. No need to
check for older version.

**Comment by carriepl** [comment]
It works with 1.8.1. Earlier versions seem to conflict with python 3.4.

**Comment by nouiz** [comment]
So it is good for me, bit someone else should double check.

**Comment by abergeron** [comment]
I don't like the test for the Gpu types, but in the current state of scan it might be better to enforce those checks.

Other than that, I'm good.

**Pull request merged**